src/widgets/table.rs

```diff
@@ -396,18 +405,28 @@ impl<'a> Table<'a> {
         self
     }
 
-    /// Get all offsets and widths of all user specified columns
-    /// Returns (x, width)
+    /// Get all offsets and widths of all user specified columns. Returns (x, width).
+    /// When self.widths is empty, it is assumed `.widths()` has not been called and a default of
+    /// equal widths is returned.
     fn get_columns_widths(&self, max_width: u16, selection_width: u16) -> Vec<(u16, u16)> {
-        let mut constraints = Vec::with_capacity(self.widths.len() * 2 + 1);
-        constraints.push(Constraint::Length(selection_width));
-        for constraint in self.widths {
-            constraints.push(*constraint);
-            constraints.push(Constraint::Length(self.column_spacing));
-        }
-        if !self.widths.is_empty() {
-            constraints.pop();
-        }
+        let widths = if self.widths.is_empty() {
+            let col_count = self
+                .rows
+                .iter()
+                .chain(self.header.iter())
+                .map(|r| r.cells.len())
+                .max()
+                .unwrap_or(0);
+            // There are `col_count - 1` spaces between the columns
+            let total_space = max_width - self.column_spacing * (col_count as u16 - 1);
```

Comment:
This panics with overflow if `max_width < column_spacing * (col_count - 1)`.
Is there something we can do to avoid the panic?
If there's not, then we need a few tests and docs to show when the panicking is expected.

Reply:
Good call! In that case, I think it makes most sense to render nothing. A panic feels unreasonable. I'll try making that a saturating subtraction instead.

Reply:
Sorry for taking so long to respond, had to duck out for a few days. As I see it, there are multiple problems to tackle here:

1. One is doing a best effort to render something if `max_width <= self.column_spacing * (col_count - 1)`. 
2. Another is deciding what to do in the case that the sum of the widths of the columns (by which I mean the maximum unicode width entry in each column) exceeds `max_width`. 

I agree that rendering the maximum number `n` columns such that `max_width > self.column_spacing * (n - 1)` is preferable to the quick fix I did with saturating subtraction for 1.

I'm not so sure about whether it's right to drop columns in 2. To the user, it mightn't be clear that there are more columns if you were to drop the ones whose full width doesn't fit the remaining space. For example, in cases like `max_width=5` and a row that looks like `("aa", "bb", "cc")`, the choice between rendering `"aa bb"`, and `"a b c"` isn't super clear. In my view, it would be helpful to have the second rendering because it's immediately apparent that I just need to minimise the font a small amount/enlarge the window to view the full table.

I think the solution I proposed to 1. would also display the second behaviour in that example, which means the code would pretty simple. What do you think? Let me know if anything isn't clear, I might not have explained super well.

I also agree that a more visual test like the one you describe would be ideal.

Reply:
If you write the visual tests how close do you think you'd get to being able to answer your questions?

There's also a reason I chose going to 15 there - for most things in software we care about how things work with 0, 1 and >1, so when truncating we want to extend that to something that truncates 0,1,>1 and leaves 0,1,>1 when truncated. So that means 3/4 for each item and 3 for the number of items. So then we look at the same values for the gaps and what happens when there isn't enough space 3*4+2(spaces)+1(selection) or 3*3+2*2(spaces)+1 etc.